samples/nrf9160/serial_lte_modem/src/tcpip_proxy/slm_at_tcp_proxy.c

```diff
@@ -0,0 +1,679 @@
+
+		if (error) {
+			sprintf(rsp_buf, "#XTCPCLI: %d disconnected\r\n",
+				error);
+			rsp_send(rsp_buf, strlen(rsp_buf));
+		}
+	}
+
+	return ret;
+}
+
+static int do_tcp_send(const u8_t *data, int datalen)
+{
+	int ret = 0;
+	u32_t offset = 0;
+	int sock;
+
+	if (proxy.role == AT_TCP_ROLE_CLIENT &&
+		proxy.sock != INVALID_SOCKET) {
+		sock = proxy.sock;
+	} else if (proxy.role == AT_TCP_ROLE_SERVER &&
+		proxy.sock_peer != INVALID_SOCKET) {
+		sock = proxy.sock_peer;
+	} else {
+		LOG_ERR("Not connected yet");
+		return -EINVAL;
+	}
+
+	while (offset < datalen) {
+		ret = send(sock, data + offset, datalen - offset, 0);
+		if (ret < 0) {
+			LOG_ERR("send() failed: %d", -errno);
+			if (errno != EAGAIN && errno != ETIMEDOUT) {
+				if (proxy.role == AT_TCP_ROLE_CLIENT) {
+					do_tcp_client_disconnect(-errno);
+				} else {
+					do_tcp_server_stop(-errno);
+				}
+			} else {
+				sprintf(rsp_buf, "#XTCPSEND: %d\r\n", -errno);
+				rsp_send(rsp_buf, strlen(rsp_buf));
+			}
+			ret = -errno;
+			break;
+		}
+		offset += ret;
+	}
+
+	sprintf(rsp_buf, "#XTCPSEND: %d\r\n", offset);
+	rsp_send(rsp_buf, strlen(rsp_buf));
+	if (ret >= 0) {
+		return 0;
+	} else {
+		return ret;
+	}
+}
+
+static void tcp_thread_func(void *p1, void *p2, void *p3)
+{
+	int ret;
+	struct pollfd fds;
+	int sock;
+
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+thread_entry:
+	if (proxy.role == AT_TCP_ROLE_SERVER) {
+		socklen_t len = sizeof(struct sockaddr_in);
+		char peer_addr[16];
+
+		/* Accept incoming connection */;
+		LOG_DBG("Accept connection...");
+		proxy.sock_peer = INVALID_SOCKET;
+		ret = accept(proxy.sock, (struct sockaddr *)&remote, &len);
+		if (ret < 0) {
+			LOG_ERR("accept() failed: %d", -errno);
+			do_tcp_server_stop(-errno);
+			return;
+		}
+		if (inet_ntop(AF_INET, &remote.sin_addr, peer_addr, len)
+			!= NULL) {
+			sprintf(rsp_buf, "#XTCPSVR: %s connected\r\n",
+				peer_addr);
+			rsp_send(rsp_buf, strlen(rsp_buf));
+		}
+		proxy.sock_peer = ret;
+		/* Start a one-shot timer to close the connection */
+		k_timer_start(&conn_timer, K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
+			0);
+	}
+
+	if (proxy.role == AT_TCP_ROLE_SERVER) {
+		sock = proxy.sock_peer;
+	} else {
+		sock = proxy.sock;
+	}
+	fds.fd = sock;
+	fds.events = POLLIN;
+	while (true) {
+		if (proxy.role == AT_TCP_ROLE_SERVER &&
+			k_timer_status_get(&conn_timer) > 0) {
+			k_timer_stop(&conn_timer);
+			LOG_INF("Connecion timeout");
+			close(proxy.sock_peer);
+			goto thread_entry;
+		}
+		ret = poll(&fds, 1, K_SECONDS(CONFIG_SLM_TCP_POLL_TIME));
+		if (ret < 0) {  /* IO error */
+			LOG_WRN("poll() error: %d", ret);
+			continue;
+		}
+		if (ret == 0) {  /* timeout */
+			continue;
+		}
+		LOG_DBG("Poll events 0x%08x", fds.revents);
+		if ((fds.revents & POLLIN) == POLLIN) {
+			char data[NET_IPV4_MTU];
+
+			ret = recv(sock, data, NET_IPV4_MTU, 0);
+			if (ret < 0) {
+				LOG_WRN("recv() error: %d", -errno);
+				continue;
+			}
+			if (ret == 0) {
+				continue;
```

Comment:
In reality these two conditions do not hit. The original idea is always to let poll() discover socket issues including remote hang-up, unfortunately not effective. Shall we keep current untouched in case we'll have poll() improvement in the future?